File: OpenMPvalidation/SRC/F90/omp_check_threadprivate.f
!********************************************************************
! Functions: check_omp_threadprivate
!********************************************************************
!Yi Wen modified this function from his own understanding of the semantics
!of C version at 05042004
!The undeestanding is that sum0 and myvalue can be local static variables
!of the check_omp_threadprivate function. There is no need to use common
!block
        integer function check_omp_threadprivate()
        implicit none
	integer sum, known_sum, i , iter, rank,size, failed
!Yi Wen modified at 05042004 : add "save"
	integer,save:: sum0
        real, save::myvalue
	integer omp_get_num_threads, omp_get_thread_num
        real my_random
	real, allocatable:: data(:)
        integer random_size
	intrinsic random_number
        intrinsic random_seed
        external omp_set_dynamic
!Yi Wen commented two common blocks
!	common/csum0/ sum0
!	common/cmyvalue/ myvalue
!!!!!!!!!!$omp threadprivate(/csum0/,/cmyvalue/)
!$omp threadprivate(sum0,myvalue)
	include "omp_testsuite.f"
	sum = 0
	failed = 0
        sum0=0
        myvalue=0
        random_size=45
        call omp_set_dynamic(0)
!$omp parallel
	sum0 = 0
!$omp do
	do i=1, LOOPCOUNT
	  sum0 = sum0 + i
	end do
!$omp end do
!$omp critical
	sum = sum + sum0
!$omp end critical
!$omp end parallel
	known_sum = (LOOPCOUNT*(LOOPCOUNT+1))/2
	if ( known_sum .ne. sum ) then
	  print *, ' known_sum =', known_sum, ', sum =',sum
	end if

	call omp_set_dynamic(0)
	
!$omp parallel
!$omp master
	size = omp_get_num_threads()
	allocate ( data(size) )
!$omp end master
!$omp end parallel
        call random_seed(SIZE=random_size)
	do iter = 0, 99
	  call random_number(harvest=my_random)
!$omp parallel private(rank)
	  rank = omp_get_thread_num()
	  myvalue = my_random + rank
	  data(rank) = myvalue
!$omp end parallel
!$omp parallel private(rank)
	  rank = omp_get_thread_num()
	  if ( myvalue .ne. data(rank) ) then
	   failed = failed + 1
	   print *, ' myvalue =',myvalue,' data(rank)=', data(rank)
	  end if
!$omp end parallel
	end do
	deallocate( data)
	if ( known_sum .eq. sum .and. failed .ne. 1) then
	  check_omp_threadprivate = 1
	else
	  check_omp_threadprivate = 0 
	end if
	end

	integer function crosscheck_omp_threadprivate()
        implicit none
	integer sum, known_sum, i, iter, rank, size, failed
	integer sum1
        real my_random
        integer, save::crosssum0
        real, save::crossmyvalue
        integer omp_get_num_threads,omp_get_thread_num
        integer random_size
	intrinsic random_number
        intrinsic random_seed
        external omp_set_dynamic
	real, allocatable :: data(:)
!	common sum0, myvalue, crossmyvalue
!Yi Wen commented comm block below at 05042004
!	common/csum1/sum1
!!!!!!!!!$omp threadprivate(/csum1/)
        include "omp_testsuite.f"
	sum1 = 789
        sum = 0
	failed = 0
        random_size=45
	call omp_set_dynamic(0)
!$omp parallel
	crosssum0 = 0
!$omp do 
	do i=1, LOOPCOUNT
	   crosssum0 = crosssum0 + i
	end do
!$omp end do
!$omp critical
	sum = sum + crosssum0
!$omp end critical
!$omp end parallel
	known_sum = (LOOPCOUNT*(LOOPCOUNT+1))/2
!$omp parallel
!$omp master
	size = omp_get_num_threads()
	allocate ( data(size) )
!$omp end master
!$omp end parallel
        call random_seed(SIZE=random_size)
	do iter = 0, 99
	   call random_number(harvest=my_random)
!$omp parallel private(rank)
	   rank = omp_get_thread_num()
	   crossmyvalue = my_random + rank
	   data(rank) = crossmyvalue
!$omp end parallel
!$omp parallel reduction(+:failed) private(rank)
	   rank = omp_get_thread_num()
	   if ( crossmyvalue .ne. data(rank) ) then
	     failed = failed + 1
 	   end if
!$omp end parallel
	end do
	deallocate(data)
	if ( known_sum .eq. sum .and. failed .ne. 1) then
	  crosscheck_omp_threadprivate = 1
	else
	  crosscheck_omp_threadprivate = 0
	end if
	end

!********************************************************************
! Functions: check_omp_copyin
!********************************************************************

	integer function check_omp_copyin()
        implicit none
	integer sum, known_sum, i
!Yi Wen at 05042004 modified to add "save" for sum1 and
!commented the below common block
        integer, save::sum1
!	common/csum1/ sum1
!!!!!!!$omp threadprivate(/csum1/)
!$omp threadprivate(sum1)
	sum = 0
	sum1 = 0
!$omp parallel copyin(sum1)
!        print *,"sum1",sum1
!$omp do
	do i=1, 999
	  sum1 = sum1 + i
	end do
!$omp critical
	sum = sum + sum1
!$omp end critical
!$omp end parallel
	known_sum = 999*1000/2
	if ( known_sum .eq. sum ) then
	   check_omp_copyin = 1
	else
	   check_omp_copyin = 0
	end if
	end

	integer function crosscheck_omp_copyin()
        implicit none
	integer sum, known_sum, i
!Yi Wen at 05042004, same modification
        integer, save::crosssum1
!	common/ccrosssum1/ crosssum1
!!!!!!!!$omp threadprivate(/ccrosssum1/)
!$omp threadprivate(crosssum1)
	sum = 0
	crosssum1 = 789
!$omp parallel
!	print *,"cs",crosssum1
!$omp do
	do i=1, 999
	  crosssum1 = crosssum1 + i
	end do
!$omp end do
!$omp critical
	sum = sum + crosssum1
!$omp end critical
!$omp end parallel
	known_sum = 999*1000/2
	if ( known_sum .eq. sum ) then
	    crosscheck_omp_copyin = 1
	else
	    crosscheck_omp_copyin = 0
	end if
	end

!********************************************************************
! Functions: check_omp_copyprivate
!********************************************************************

	integer function check_omp_copyprivate()
        implicit none
	integer sum, known_sum, i, sum0
	common /csum0/ sum0
!$omp threadprivate(/csum0/)
	sum = 0
!$omp parallel
	do i=1, 999
!$omp single 
	  sum0 = sum0 + i
!$omp end single copyprivate(/csum0/)
!$omp critical
	  sum = sum + sum0
!$omp end critical
	end do
!$omp end parallel
	known_sum = 999* 1000/2
	if ( known_sum .eq. sum ) then
	   check_omp_copyprivate = 1
	else
	   check_omp_copyprivate = 0
	end if
	end

	integer function crosscheck_spmd_threadprivate()
!        implicit none
	integer iter, rank, size, failed 
        real myvalue2
        real my_random
        integer random_size
	intrinsic random_number
        intrinsic random_seed
	real, allocatable:: data(:)
	integer omp_get_num_threads, omp_get_thread_num
	failed = 0
	myvalue2 = 2
        random_size=45
!$omp parallel
!$omp master
	size = omp_get_num_threads()
	allocate( data(size) )
!$omp end master
!$omp end parallel
	do iter=0, 999
!$omp parallel private(rank)
	   rank = omp_get_thread_num()
           call random_seed(SIZE=random_size)
	   call random_number(harvest=my_random)
	   myvalue2 = my_random + rank
!$omp end parallel
!$omp parallel reduction(+:failed) private(rank)
	   rank = omp_get_thread_num()
	   if ( myvalue2 .ne. data(rank) ) then
	     failed = failed + 1
	   end if
!$omp end parallel
	end do
	deallocate ( data )
	if ( failed .ne. 1) then
	   crosscheck_spmd_threadprivate = 1
	else
	   crosscheck_spmd_threadprivate = 0
	end if
	end


!***************************************************************************
! driver function
!***************************************************************************

      subroutine omp_check_threadprivate(N,failed,num_tests,
     x  crosschecked)
		implicit none
		integer, external::check_omp_threadprivate
		integer, external::crosscheck_omp_threadprivate
                integer, external::check_omp_copyin
                integer, external::crosscheck_omp_copyin
		character (len=30)::name
		integer failed
		integer N,num_tests,crosschecked
		name="OMP THREADPRIVATE"
		call do_test(check_omp_threadprivate,
     x      crosscheck_omp_threadprivate,name,N,failed,
     x      num_tests,crosschecked)
                name="OMP COPYIN"
	call do_test(check_omp_copyin,crosscheck_omp_copyin,
     x  name,N,failed,num_tests,crosschecked)
		end

File: OpenMPvalidation/SRC/C/omp_check_threadprivate.c

#include "omp_testsuite.h"
#include <stdlib.h>
#include <stdio.h>
#include <omp.h>

static int sum0=0;
#pragma omp threadprivate(sum0)

static int myvalue=0;
#pragma omp threadprivate(myvalue)

int check_omp_threadprivate()
{
  int sum=0;
  int known_sum;
  int i;

  int iter;

  int *data;
  int size;
  int failed=0;
  int my_random;
  omp_set_dynamic(0);

#pragma omp parallel 
 {
   sum0=0;
#pragma omp for 
  for (i=1;i<=LOOPCOUNT;i++)
    {
      sum0=sum0+i;
    }                       /*end of for*/
#pragma omp critical
  {
    sum= sum+sum0;
  }                         /*end of critical*/
 }                          /* end of parallel*/    
  known_sum=(LOOPCOUNT*(LOOPCOUNT+1))/2;
  if (known_sum != sum ) {
    printf(" known_sum = %d , sum = %d \n",known_sum,sum);
  }


  /* the next parallel region is just used to get the number of threads*/
  omp_set_dynamic(0);
#pragma omp parallel
 {
   #pragma omp master
   {
     size=omp_get_num_threads();
     data=(int*) malloc(size*sizeof(int));
   }
 }/* end parallel*/

 
  srand(45);
  for (iter=0; iter<100; iter++){
    my_random=rand();/* random number generator is called inside serial region*/

    /* the first parallel region is used to initialiye myvalue and the array with my_random+rank*/
#pragma omp parallel
    {
	int rank;
	rank=omp_get_thread_num();
	myvalue=data[rank]=my_random+rank;
    }

    /* the second parallel region verifies that the value of "myvalue" is retained */
#pragma omp parallel reduction(+:failed)
    {
	int rank;
	rank=omp_get_thread_num();
	failed = failed + (myvalue != data[rank]);
	if(myvalue != data[rank]){
	    printf(" myvalue = %d, data[rank]= %d\n",myvalue,data[rank]);
      }
    }
  }
  free (data);

  return (known_sum==sum) && !failed;

}/* end of check_threadprivate*/




static int crosssum0=0;
/*#pragma omp threadprivate(crosssum0)*/

static int crossmyvalue=0;
/*#pragma omp threadprivate(crossmyvalue)*/

int crosscheck_omp_threadprivate()
{
  int sum=0;
  int known_sum;
  int i;

  int iter;

  int *data;
  int size;
  int failed=0;
  int my_random;
  omp_set_dynamic(0);

#pragma omp parallel 
 {
   crosssum0=0;
#pragma omp for 
  for (i=1;i<LOOPCOUNT;i++)
    {
      crosssum0=crosssum0+i;
    }                       /*end of for*/
#pragma omp critical
  {
    sum= sum+crosssum0;
  }                         /*end of critical*/
 }                          /* end of parallel*/    
  known_sum=(LOOPCOUNT*(LOOPCOUNT+1))/2;



#pragma omp parallel
 {
   #pragma omp master
   {
     size=omp_get_num_threads();
     data=(int*) malloc(size*sizeof(int));
   }
 }
  srand(45);
  for (iter=0; iter<100; iter++){
    my_random=rand();
#pragma omp parallel
    {
	int rank;
	rank=omp_get_thread_num();
	crossmyvalue=data[rank]=my_random+rank;
    }
#pragma omp parallel reduction(+:failed)
    {
	int rank;
	rank=omp_get_thread_num();
	failed = failed + (crossmyvalue != data[rank]);
    }
  }
  free (data);

  return (known_sum==sum) && !failed;

}/* end of crosscheck_threadprivate*/






static int sum1=789;
#pragma omp threadprivate(sum1)

int check_omp_copyin()
{
  int sum=0;
  int known_sum;
  int i;
  sum1=0;
#pragma omp parallel copyin(sum1)
 {
   /*printf("sum1=%d\n",sum1);*/
#pragma omp for 
  for (i=1;i<1000;i++)
    {
      sum1=sum1+i;
    }                       /*end of for*/
#pragma omp critical
  {
    sum= sum+sum1;
  }                         /*end of critical*/
 }                          /* end of parallel*/    
  known_sum=(999*1000)/2;
  return (known_sum==sum);

}/* end of check_threadprivate*/

static int crosssum1=789;
#pragma omp threadprivate(crosssum1)

int crosscheck_omp_copyin()
{
  int sum=0;
  int known_sum;
  int i;
  crosssum1=0;
#pragma omp parallel
 {
   /*printf("sum1=%d\n",sum1);*/
#pragma omp for 
  for (i=1;i<1000;i++)
    {
      crosssum1=crosssum1+i;
    }                       /*end of for*/
#pragma omp critical
  {
    sum= sum+crosssum1;
  }                         /*end of critical*/
 }                          /* end of parallel*/    
  known_sum=(999*1000)/2;
  return (known_sum==sum);

}/* end of check_threadprivate*/



/*#pragma omp threadprivate(sum0)
int check_omp_copyprivate()
{
  int sum=0;
  int known_sum;
  int i;
#pragma omp parallel 
  {
    for (i=1;i<1000;i++)
      {
#pragma omp single copyprivate(sum0) 
	{
	  sum0=sum0+i;
	} 
      

#pragma omp critical
    {
      sum= sum+sum0;
    }                        
}                     
      }                   
  known_sum=(999*1000)/2;
  return (known_sum==sum);
  
}*/




static int myvalue2=0;


int crosscheck_spmd_threadprivate(){
  int iter;

  int *data;
  int size;
  int failed=0;
  int my_random;
  omp_set_dynamic(0);
  
#pragma omp parallel
 {
   #pragma omp master
   {
     size=omp_get_num_threads();
     data=(int*) malloc(size*sizeof(int));
   }
 }
  srand(45);
  for (iter=0; iter<100; iter++){
    my_random=rand();
#pragma omp parallel
    {
	int rank;
	rank=omp_get_thread_num();
	myvalue2=data[rank]=my_random+rank;
    }
#pragma omp parallel reduction(+:failed)
    {
	int rank;
	rank=omp_get_thread_num();
	failed = failed + (myvalue2 != data[rank]);
    }
  }
  free (data);
  return !failed;
}




